src/Cli/dotnet/commands/dotnet-workload/install/MsiInstallerBase.cs

```diff
@@ -487,5 +492,75 @@ protected void UpdateDependent(InstallRequestType requestType, string providerKe
                 ExitOnFailure(response, $"Failed to update dependent, providerKey: {providerKeyName}, dependent: {dependent}.");
             }
         }
+
+        /// <summary>
+        /// Deletes the specified install state file.
+        /// </summary>
+        /// <param name="path">The path of the install state file to delete.</param>
+        protected void RemoveInstallStateFile(string path)
+        {
+            VerifyInstallStateFile(path);
+            if (File.Exists(path))
+            {
+                Log?.LogMessage($"Install state file does not exist: {path}");
+                return;
+            }
+
+            Elevate();
+
+            if (IsElevated)
+            {
+                File.Delete(path);
+            }
+            else if (IsClient)
+            {
+                InstallResponseMessage response = Dispatcher.SendRemoveInstallStateFileRequest(path);
+                ExitOnFailure(response, $"Failed to remove install state file: {path}");
+            }
+        }
+
+        /// <summary>
+        /// Verify that the specified path points to an install state file.
+        /// </summary>
+        /// <exception cref="ArgumentException" />
+        /// <remarks>
+        /// This method is intended to avoid arbitrary file creation and delete operation when executing install state
+        /// file related operations.
+        /// </remarks>
+        private void VerifyInstallStateFile(string path)
+        {
+            if (string.IsNullOrWhiteSpace(path) ||
+                (!path.StartsWith(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData)) &&
+                !path.EndsWith("default.json")))
+            {
+                throw new ArgumentException(string.Format(LocalizableStrings.InvalidInstallStateFile, path));
+            }
+        }
+
+        /// <summary>
+        /// Writes the contents of the install state JSON file.
+        /// </summary>
+        /// <param name="path">The path of the isntall state file to write.</param>
+        /// <param name="jsonLines">The contents of the JSON file, formatted as a single line.</param>
+        protected void WriteInstallStateFile(string path, IEnumerable<string> jsonLines)
+        {
+            VerifyInstallStateFile(path);
+            Elevate();
+
+            if (IsElevated)
+            {
+                // Create the parent folder for the state file and set up all required ACLs
+                MsiPackageCache.CreateSecureDirectory(Path.GetDirectoryName(path));
+
+                File.WriteAllLines(path, jsonLines);
+
+                MsiPackageCache.SecureFile(path);
```

Comment:
In this case `CreateSecureDirectory` and `SecureFile` don't seem to be related to the MSI Package Cache.  Should we rename the `MsiPackageCache` class or refactor it somehow?

Reply:
Good point, we can factor out the generic methods and leave just the cache specific pieces in this class.